server/services/color_utils.py

```
@@ -0,0 +1,118 @@
+    # Fallback: return the original lowercased
+    return [lower]
+
+
+def hex_to_hsl(hex_color: str) -> tuple[float, float, float]:
+    """Convert hex color to HSL. Returns (h, s, l) where h is 0-360, s and l are 0-1.
+
+    Raises ValueError for malformed hex strings.
+    """
+    hex_color = hex_color.lstrip("#")
+    if len(hex_color) != 6:
+        raise ValueError(f"Expected 6-digit hex color, got '{hex_color}'")
```

Ruff (0.15.5)

[warning] Avoid specifying long messages outside the exception class

(TRY003)